flink/src/main/java/org/apache/iceberg/flink/FlinkTypeToType.java

```diff
@@ -0,0 +1,149 @@
+
+package org.apache.iceberg.flink;
+
+import java.util.List;
+import org.apache.flink.table.types.AtomicDataType;
+import org.apache.flink.table.types.CollectionDataType;
+import org.apache.flink.table.types.FieldsDataType;
+import org.apache.flink.table.types.KeyValueDataType;
+import org.apache.flink.table.types.logical.BigIntType;
+import org.apache.flink.table.types.logical.BinaryType;
+import org.apache.flink.table.types.logical.BooleanType;
+import org.apache.flink.table.types.logical.CharType;
+import org.apache.flink.table.types.logical.DateType;
+import org.apache.flink.table.types.logical.DecimalType;
+import org.apache.flink.table.types.logical.DoubleType;
+import org.apache.flink.table.types.logical.FloatType;
+import org.apache.flink.table.types.logical.IntType;
+import org.apache.flink.table.types.logical.LocalZonedTimestampType;
+import org.apache.flink.table.types.logical.LogicalType;
+import org.apache.flink.table.types.logical.RowType;
+import org.apache.flink.table.types.logical.SmallIntType;
+import org.apache.flink.table.types.logical.TimeType;
+import org.apache.flink.table.types.logical.TimestampType;
+import org.apache.flink.table.types.logical.TinyIntType;
+import org.apache.flink.table.types.logical.VarBinaryType;
+import org.apache.flink.table.types.logical.VarCharType;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.types.Type;
+import org.apache.iceberg.types.Types;
+
+public class FlinkTypeToType extends FlinkTypeVisitor<Type> {
+  private final FieldsDataType root;
+  private int nextId = 0;
+
+  FlinkTypeToType(FieldsDataType root) {
+    this.root = root;
+    // the root struct's fields use the first ids
+    this.nextId = root.getFieldDataTypes().size();
+  }
+
+  private int getNextId() {
+    int next = nextId;
+    nextId += 1;
+    return next;
+  }
+
+  @Override
+  public Type fields(FieldsDataType fields, List<Type> types) {
+    List<Types.NestedField> newFields = Lists.newArrayListWithExpectedSize(types.size());
+    boolean isRoot = root == fields;
+
+    List<RowType.RowField> rowFields = ((RowType) fields.getLogicalType()).getFields();
+    Preconditions.checkArgument(rowFields.size() == types.size(), "fields list and types list should have same size.");
+
+    for (int i = 0; i < rowFields.size(); i++) {
+      int id = isRoot ? i : getNextId();
+
+      RowType.RowField field = rowFields.get(i);
+      String name = field.getName();
+      String comment = field.getDescription().orElse(null);
+
+      if (field.getType().isNullable()) {
+        newFields.add(Types.NestedField.optional(id, name, types.get(i), comment));
+      } else {
+        newFields.add(Types.NestedField.required(id, name, types.get(i), comment));
+      }
+    }
+
+    return Types.StructType.of(newFields);
+  }
+
+  @Override
+  public Type collection(CollectionDataType collection, Type elementType) {
```

Comment:
A `CollectionDataType` may be `MultisetType` too. Maybe iceberg not support it now? Or we can map it to `Map<T, Integer>`.